src/setup-beam.js

```diff
@@ -592,6 +596,327 @@ function debugLog(groupName, message) {
   )
 }
 
+function hexMirrorsInput() {
+  return core.getMultilineInput('hexpm-mirrors', {
+    required: false,
+  })
+}
+
+async function doWithMirrors(opts) {
+  const { hexMirrors, actionTitle, action } = opts
+  let actionRes
+
+  if (hexMirrors.length === 0) {
+    throw new Error(`Could not ${actionTitle} from any hex.pm mirror`)
+  }
+
+  const [hexMirror, ...hexMirrorsT] = hexMirrors
+  try {
+    actionRes = await action(hexMirror)
+  } catch (err) {
+    core.info(
+      `Action ${actionTitle} failed for mirror ${hexMirror}, with ${err}`,
+    )
+    core.debug(`Stacktrace: ${err.stack}`)
+    actionRes = await doWithMirrors({
+      hexMirrors: hexMirrorsT,
+      actionTitle,
+      action,
+    })
+  }
+
+  return actionRes
+}
+
+async function install(toolName, opts) {
+  const { osVersion, toolVersion, hexMirror } = opts
+  const versionSpec =
+    osVersion !== undefined ? `${osVersion}/${toolVersion}` : toolVersion
+  let installOpts
+
+  // The installOpts object is composed of supported processPlatform keys
+  // (e.g. 'linux', 'win32', or 'all' - in case there's no distinction between platforms)
+  // In each of these keys there's an object with keys:
+  // * downloadToolURL
+  //     - where to fetch the downloadable from
+  // * extract
+  //     - if the downloadable is compressed: how to extract it
+  //       - return ['dir', targetDir]
+  //     - if the downloadable is not compressed: a filename.ext you want to cache it under
+  //       - return ['file', filenameWithExt]
+  // * postExtract
+  //     - stuff to execute outside the cache scope (just after it's created)
+  // * reportVersion
+  ///    - configuration elements on how to output the tool version, post-install
```

Comment:
This (in a gist) explains the major change behind this pull request.

If it's not clear (since this is supposed to explain the "behaviour" required for adding new `tool-cache`d stuff) it can be improved (variables renamed, behaviour made more explicit, etc.).